### PR TITLE
the subs not error but could not start the api apolo server

### PR DESCRIPTION
**Pull request by aldoprogrammer**
## Related Issue(s)

Fixes # For #

## Proposed Changes

- [x] feat:
- [x] fix:

## Checklist

- [x] I have followed the repository's contribution guidelines.
- [x] I have added necessary documentation or updated existing documentation.
- [x] My code follows the established code style and formatting of the project.
- [x] I have tested my changes and they function as expected.
- [x] I have added suitable unit tests (if applicable).

## To-Do

1.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,8 +17,8 @@
     "gen-typings": "ts-node src/generate-typings",
     "gen-typings:watch": "ts-node src/generate-typings --watch",
     "start": "nest start",
-    "start:dev": "MODE=dev nest start --watch -b swc --type-check",
-    "start:dev:agile": "MODE=dev nest start --watch",
+    "start:dev": " nest start --watch -b swc --type-check",
+    "start:dev:agile": " nest start --watch",
```
Comment (Nasfame) [inline package.json:21]: don't remove MODE=dev
Use a separate make file locally with `npx:` this is the flow start:dev is only possible MODE=dev

File: src/app/app.module.ts [hunk 2]
```diff
@@ -99,7 +99,7 @@ console.debug({ mode, envFile })
     CacheModule.register({ isGlobal: true }),
     GraphQLModule.forRoot<ApolloDriverConfig>({
       context: ({ req, res, connection }: GraphqlContext) => {
-        /*if (connection) { //DOesn't work
+        if (connection) { //DOesn't work
```
Comment (Nasfame) [inline src/app/app.module.ts:102]: is this working now???

File: src/app/app.module.ts [hunk 3]
```diff
@@ -123,9 +123,9 @@ console.debug({ mode, envFile })
           path: '/graphql',
         },
         // 'graphql-ws': true,
-        'subscriptions-transport-ws': true,
+        // 'subscriptions-transport-ws': true,
 
-        /* 'subscriptions-transport-ws': {
+         'subscriptions-transport-ws': {
```
Comment (Nasfame) [inline src/app/app.module.ts:128]: is this working as exxpected

**Review by Nasfame — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (aldoprogrammer) [reply package.json:21]: I just remove it cs i run it on windows

File: src/app/app.module.ts (hunk 2 above)
Comment (aldoprogrammer) [reply src/app/app.module.ts:102]: Waitt a min

Comment (aldoprogrammer) [reply src/app/app.module.ts:102]: I'll make a change in my code in local lmachine

Comment (Nasfame) [reply src/app/app.module.ts:102]: ok

File: src/app/app.module.ts (hunk 3 above)
Comment (aldoprogrammer) [reply src/app/app.module.ts:128]: nothing works, on another side it also said 
```
[start:dev:agile] > MODE=dev nest start --watch
[start:dev:agile]
[start:dev:agile] Invalid parameter - =dev
``` if I add the mode dev, but when I remove them, then the error is gone but when i click on http://localhost:3000 or http://localhost:3000/graphql it says site can't be reached
![nih](https://github.com/learnery-engine/api.creator.learnery/assets/107610584/644fd9db-05a6-421a-9b02-90f09ec9ad64)
![Uploading apaaa.JPG…]()

Comment (aldoprogrammer) [reply src/app/app.module.ts:128]: `[start:dev:agile] { mode: undefined, envFile: undefined }` it gives this too actually, is that means wrong or smth?

Comment (Nasfame) [reply src/app/app.module.ts:128]: U r using windows that's why its throwing up the error. use `set MODE=development` ; and do npm run start. 

or write a custom script for `start:dev:agile:windows` . Don't change unnecessary things. You should realize other team members are working on it and you should break their development.

**Comment by Nasfame** [comment]
@aldoprogrammer i have updated ur branch. So do
```
git stash
git pull
git stash pop
```

File: src/app/app.module.ts (hunk 3 above)
Comment (aldoprogrammer) [reply src/app/app.module.ts:128]: Oh sorry, I didn't know that

Comment (aldoprogrammer) [reply src/app/app.module.ts:128]: I was following this `start:dev:agile:windows` but still the same result.

**Comment by aldoprogrammer** [comment]
@Nasfame when will the zoom meeting? maybe i'd like to join once in a while

**Comment by aldoprogrammer** [comment]
> @aldoprogrammer i have updated ur branch. So do
> 
> ```
> git stash
> git pull
> git stash pop
> ```

**Comment by aldoprogrammer** [comment]
```
git stash
git pull
git stash pop
```

this command does not work on my machine. do you mean to make another pr to this repo branch?

**Comment by aldoprogrammer** [comment]
@Nasfame oh my goddness, im so sorry, i didn't notice that the server is running on port 1606,not 3000. that's really my bad

**Comment by aldoprogrammer** [comment]
now i can see the playground, but not the apollo server 
image

No error found, and here is the app.modul

```import {
  INestApplication,
  MiddlewareConsumer,
  Module,
  NestModule,
  OnModuleInit,
} from '@nestjs/common'
import { AppService } from './app.service'
import { ConfigModule, ConfigService } from '@nestjs/config'
import { AuthModule } from '../auth/auth.module'
import Joi from 'joi'
import { CacheModule } from '@nestjs/cache-manager'
import { GraphQLModule } from '@nestjs/graphql'
import { ApolloServerPluginLandingPageGraphQLPlayground } from '@apollo/server-plugin-landing-page-graphql-playground'
import { GraphQLDateTime } from 'graphql-iso-date'
import { ApolloDriver, ApolloDriverConfig } from '@nestjs/apollo'
import { AppController } from './app.controller'
import { GraphqlContext } from './dto/request.dto'
import { UserModule } from '../user/user.module'
import { PrismaModule } from '../prisma/prisma.module'
import { CategoryController } from '../category/category.controller'
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger'
import { CategoryModule } from '../category/category.module'
import { UserController } from '../user/user.controller'
import { AuthController } from '../auth/auth.controller'
import { CategoryService } from '../category/category.service'
import { UserService } from '../user/user.service'
import { AuthService } from '../auth/auth.service'
import { JwtService } from '@nestjs/jwt'
import { ClerkModule } from '../clerk/clerk.module'
import { ClerkService } from '../clerk/clerk.service'
import { CourseModule } from '../course/course.module'
import { HttpLoggerMiddleware } from './middlewares/http-logger.middleware'
import { AiModule } from '../ai/ai.module'
import { AiService } from '../ai/ai.service'
import { PineconeService } from '@src/ai/pinecone/pinecone.service'
import { OpenAIService } from '@src/ai/openai/openAIService'

let mode = process.env.MODE
let envFile: string

let cookieName: string

switch (mode) {
  case 'test':
    envFile = '.env.test'
    process.env.NODE_ENV = 'testing'
    break
  case 'prod':
    process.env.NODE_ENV = 'production'
    envFile = '.env.prod'
    break
  case 'dev':
    mode = 'dev'
    process.env.NODE_ENV = 'development'
    envFile = '.env.local'
    break
}

console.debug({ mode, envFile })

@Module({
  imports: [
    ConfigModule.forRoot({
      envFilePath: [envFile, '.env.local'], //by default looks for .env file in root directory
      isGlobal: true,
      cache: true,
      validationSchema: Joi.object({
        MODE: Joi.string().valid('dev', 'prod', 'test').default('dev'),
        PORT: Joi.number().default(1606),
        DATABASE_URL: Joi.string(),
        JWT_SECRET: Joi.string().default('Hiro@JWT#TOKEN$'),
        JWT_MAX_AGE: Joi.number()
          .default(Number.MAX_SAFE_INTEGER >> 1) //2 * 60 * 60 * 1000)
          .max(Number.MAX_SAFE_INTEGER - 1) //to round to 0
          .min(60 * 1000),
        CLERK_SECRET_KEY: Joi.string().required().min(10),
        DOMAIN: Joi.string().optional().default(''),
        COOKIE_PATH: Joi.string().optional().default('/').min(1),
        COOKIE_NAME: Joi.string().optional().default('token').min(2),
        OPENAI_API_KEY: Joi.string().required().min(20), //FIXME: change length
        OPENAI_CHAT_MODEL: Joi.string()
          .optional()
          .default('gpt-3.5-turbo-16k')
          .valid(
            'gpt-3.5-turbo',
            'gpt-3.5-turbo-0613',
            'gpt-3.5-turbo-16k-0613',
            'gpt-3.5-turbo-16k',
            'gpt-4',
            'gpt-4-32k',
            'gpt-3.5-turbo-16k-0613',
          ),
        PINECONE_API_KEY: Joi.string().required().min(30),
        PINECONE_API_ENV: Joi.string().required().min(7),
        PINECONE_INDEX: Joi.string().optional().default('learnery'),
      }),
    }),
    CacheModule.register({ isGlobal: true }),
    GraphQLModule.forRoot<ApolloDriverConfig>({
      context: ({ req, res, connection }: GraphqlContext) => {
        if (connection) { //DOesn't work
          // This is a WebSocket connection
          console.log({connection})
          return { req: connection.context.req, res: connection.context.res };
        }
        // This is an HTTP request*/
        return { req, res }
      }, //required for cookies
      introspection: true, //mode!='prod',//FIXME: disable in production
      fieldResolverEnhancers: ['interceptors'], //FIXME:unsure
      installSubscriptionHandlers: true,

      // https://docs.escape.tech/cookbooks/enabling-introspection#:~:text=The%20package%20is%20a%20wrapper,in%20the%20NestJS%20related%20documentation.&text=By%20default%2C%20the%20introspection%20is,true%20in%20the%20module%20declaration.
      autoSchemaFile: false, // schema-first
      playground: false,
      driver: ApolloDriver,
      plugins: [ApolloServerPluginLandingPageGraphQLPlayground()],
      typePaths: ['./**/*.graphql'],
      resolvers: { DateTime: GraphQLDateTime },
      subscriptions: {
        'graphql-ws': {
          path: '/graphql',
        },
        // 'graphql-ws': true,
        // 'subscriptions-transport-ws': true,

         'subscriptions-transport-ws': {
          onConnect: (
            connectionParams: any,
            websocket: any,
            ctx: any,
          ): GraphqlContext => {
            console.log({ req: ctx.req });
            return { req: ctx.req, res: ctx.res };
            // console.log({ req: ctx.connection.context.req })
            return { req: ctx.connection.context.req, res: ctx.connection.context.res }
          },
        },
      },
    }),
    PrismaModule,
    AuthModule,
    UserModule,
    CategoryModule,
    ClerkModule,
    CourseModule,
    AiModule,
  ],
  controllers: [AppController, CategoryController, UserController, AuthController],
  providers: [
    /*  {
      provide: APP_FILTER, //FIXME; not working try with invalid pinecone cred
      useClass: GlobalExceptionFilter,
    },*/
    AppService,
    CategoryService,
    UserService,
    AuthService,
    JwtService,
    ClerkService,
    OpenAIService,
    PineconeService,
    AiService,
  ],
})
export class AppModule implements OnModuleInit, NestModule {
  configure(consumer: MiddlewareConsumer): void {
    consumer.apply(HttpLoggerMiddleware).forRoutes('*')
  }
  constructor(private readonly config: ConfigService) {
    /*cookieName = config.get('COOKIE_NAME')
    console.debug({"SET-COOKIE": cookieName})*/
  }

  onModuleInit() {
    cookieName = this.config.get('COOKIE_NAME')
    console.debug({ 'SET-COOKIE': cookieName })
  }

  static createDocument(app: INestApplication) {
    const options = new DocumentBuilder()
      .setTitle('Learnery Creator APIs')
      .setDescription('apis to crud courses')
      .setVersion('1.0')
      .addTag('learnery-creator')
      .addBearerAuth()
      .addCookieAuth(cookieName)
      .build()
    return SwaggerModule.createDocument(app, options)
  }
}
```

**Comment by aldoprogrammer** [comment]
im testing on playground and got this `"message": "Cannot read properties of undefined (reading 'userId')"` i guess this is bcs of the auth guard right? i already tried to remove the guard still find somme error, maybe you can teach me a little bit about this so that i can solve it fastly

**Comment by Nasfame** [comment]
@aldoprogrammer u can hold this pr for now. I am assigning u a new task check jira

File: src/app/app.module.ts (hunk 3 above)
Comment (Nasfame) [reply src/app/app.module.ts:128]: > nothing works, on another side it also said
> 
> ```
> [start:dev:agile] > MODE=dev nest start --watch
> [start:dev:agile]
> [start:dev:agile] Invalid parameter - =dev
> ```
> 
> if I add the mode dev, but when I remove them, then the error is gone but when i click on http://localhost:3000 or http://localhost:3000/graphql it says site can't be reached apaaa nih

the port is 1606

Comment (Nasfame) [reply src/app/app.module.ts:128]: checkout localhost:1606/graphql

Comment (Nasfame) [reply src/app/app.module.ts:128]: always make sure the port is correct @aldoprogrammer

Comment (Nasfame) [reply src/app/app.module.ts:128]: if u want port to be 3000 specify it in .env

**Comment by Nasfame** [comment]
soft fix in #51

**Pull request closed without merging**